File: ydb/public/sdk/python/examples/pagination/pagination.py
# -*- coding: utf-8 -*-
from __future__ import print_function
import os

import ydb
from concurrent.futures import TimeoutError
from . import sample_data


class SchoolsPaginated(object):
    def __init__(self, session, path, limit):
        self.session = session
        self.path = path
        self.limit = limit
        self.last_city = None
        self.last_number = None

        next_page_query = """
        PRAGMA TablePathPrefix("{path}");

        DECLARE $limit AS Uint32;
        DECLARE $lastCity AS Utf8;
        DECLARE $lastNumber AS Uint32;

        $Part1 = (
            SELECT * FROM schools
            WHERE `city` = $lastCity AND `number` > $lastNumber
            ORDER BY `city`, `number`
            LIMIT $limit
        );

        $Part2 = (
            SELECT * FROM schools
            WHERE `city` > $lastCity
            ORDER BY `city`, `number`
            LIMIT $limit
        );

        $Union = (SELECT * FROM $Part1 UNION ALL SELECT * FROM $Part2);
        SELECT * FROM $Union ORDER BY `city`, `number` LIMIT $limit;
        """.format(
            path=self.path
        )

        self.prepared_next_page_query = self.session.prepare(next_page_query)

    def get_first_page(self):
        query = """
        PRAGMA TablePathPrefix("{path}");

        DECLARE $limit AS Uint32;

        SELECT * FROM schools
        ORDER BY `city`, `number`
        LIMIT $limit;
        """.format(
            path=self.path
        )
        prepared_query = self.session.prepare(query)
        result_sets = self.session.transaction(ydb.SerializableReadWrite()).execute(
            prepared_query, {"$limit": self.limit}, commit_tx=True
        )
        return result_sets[0]

    def get_next_page(self):
        result_sets = self.session.transaction(ydb.SerializableReadWrite()).execute(
            self.prepared_next_page_query,
            {
                "$limit": self.limit,
                "$lastCity": self.last_city,
                "$lastNumber": self.last_number,
            },
            commit_tx=True,
        )
        return result_sets[0]

    def pages(self):
        while True:
            if self.last_city is None or self.last_number is None:
                result = self.get_first_page()
            else:
                result = self.get_next_page()
            if not result.rows:
                return
            last_row = result.rows[-1]
            self.last_city = last_row.city
            self.last_number = last_row.number
            yield result


def run(endpoint, database, path, auth_token):
    driver_config = ydb.DriverConfig(endpoint, database=database, auth_token=auth_token)
    try:
        driver = ydb.Driver(driver_config)
        driver.wait(timeout=5)
    except TimeoutError:
        raise RuntimeError("Connect failed to YDB")

    try:
        session = driver.table_client.session().create()
        sample_data.prepare_data(driver, session, database, path)
        db_path = os.path.join(database, path)

        page_num = 0
        for page in SchoolsPaginated(session, db_path, limit=3).pages():
            page_num = page_num + 1
            print("-- page: {} --".format(page_num))
            for row in page.rows:
                print(row.city.encode("utf-8"), row.number, row.address.encode("utf-8"))

    finally:
        driver.stop()

File: ydb/public/sdk/cpp/examples/pagination/pagination.cpp
#include "pagination.h"

#include <util/string/cast.h>

#include <filesystem>
#include <format>

using namespace NYdb;
using namespace NYdb::NTable;
using namespace NYdb::NStatusHelpers;

const uint32_t MaxPages = 10;

static std::string JoinPath(const std::string& basePath, const std::string& path) {
    if (basePath.empty()) {
        return path;
    }

    std::filesystem::path prefixPathSplit(basePath);
    prefixPathSplit /= path;

    return prefixPathSplit;
}

//! Creates sample table with CrateTable API.
static void CreateTable(TTableClient client, const std::string& path) {
    ThrowOnError(client.RetryOperationSync([path](TSession session) {
        auto schoolsDesc = TTableBuilder()
            .AddNullableColumn("city", EPrimitiveType::Utf8)
            .AddNullableColumn("number", EPrimitiveType::Uint32)
            .AddNullableColumn("address", EPrimitiveType::Utf8)
            .SetPrimaryKeyColumns({ "city", "number" })
            .Build();

        return session.CreateTable(JoinPath(path, "schools"), std::move(schoolsDesc)).GetValueSync();
    }));
}

//! Fills sample tables with data in single parameterized data query.
static TStatus FillTableDataTransaction(TSession& session, const std::string& path) {
    auto query = std::format(R"(
        --!syntax_v1
        PRAGMA TablePathPrefix("{}");

        DECLARE $schoolsData AS List<Struct<
            city: Utf8,
            number: Uint32,
            address: Utf8>>;

        REPLACE INTO schools
        SELECT
            city,
            number,
            address
        FROM AS_TABLE($schoolsData);
    )", path);

    auto params = GetTablesDataParams();

    return session.ExecuteDataQuery(
        query,
        TTxControl::BeginTx(TTxSettings::SerializableRW()).CommitTx(),
        std::move(params)).GetValueSync();
}

//! Shows usage of query paging.
static TStatus SelectPagingTransaction(TSession& session, const std::string& path,
    uint64_t pageLimit, const std::string& lastCity, uint32_t lastNumber, std::optional<TResultSet>& resultSet)
{
    auto query = std::format(R"(
        --!syntax_v1
        PRAGMA TablePathPrefix("{}");

        DECLARE $limit AS Uint64;
        DECLARE $lastCity AS Utf8;
        DECLARE $lastNumber AS Uint32;

        $part1 = (
            SELECT * FROM schools
            WHERE city = $lastCity AND number > $lastNumber
            ORDER BY city, number LIMIT $limit
        );

        $part2 = (
            SELECT * FROM schools
            WHERE city > $lastCity
            ORDER BY city, number LIMIT $limit
        );

        $union = (
            SELECT * FROM $part1
            UNION ALL
            SELECT * FROM $part2
        );

        SELECT * FROM $union
        ORDER BY city, number LIMIT $limit;
    )", path);

    auto params = session.GetParamsBuilder()
        .AddParam("$limit")
            .Uint64(pageLimit)
            .Build()
        .AddParam("$lastCity")
            .Utf8(lastCity)
            .Build()
        .AddParam("$lastNumber")
            .Uint32(lastNumber)
            .Build()
        .Build();

    auto result = session.ExecuteDataQuery(
        query,
        TTxControl::BeginTx(TTxSettings::SerializableRW()).CommitTx(),
        std::move(params)).GetValueSync();

    if (result.IsSuccess()) {
        resultSet = result.GetResultSet(0);
    }

    return result;
}

bool SelectPaging(TTableClient client, const std::string& path, uint64_t pageLimit, std::string& lastCity, uint32_t& lastNumber) {
    std::optional<TResultSet> resultSet;
    ThrowOnError(client.RetryOperationSync([path, pageLimit, &lastCity, lastNumber, &resultSet](TSession session) {
        return SelectPagingTransaction(session, path, pageLimit, lastCity, lastNumber, resultSet);
    }));

    TResultSetParser parser(*resultSet);

    if (!parser.TryNextRow()) {
        return false;
    }
    do {
        lastCity = parser.ColumnParser("city").GetOptionalUtf8().value();
        lastNumber = parser.ColumnParser("number").GetOptionalUint32().value();
        std::cout << lastCity << ", Школа №" << lastNumber << ", Адрес: " << parser.ColumnParser("address").GetOptionalUtf8().value_or("(NULL)") << std::endl;
    } while (parser.TryNextRow());
    return true;
}

bool Run(const TDriver& driver, const std::string& path) {
    TTableClient client(driver);

    try {
        CreateTable(client, path);

        ThrowOnError(client.RetryOperationSync([path](TSession session) {
            return FillTableDataTransaction(session, path);
        }));

        uint64_t limit = 3;
        std::string lastCity;
        uint32_t lastNumber = 0;
        uint32_t page = 0;
        bool pageNotEmpty = true;

        std::cout << "> Pagination, Limit=" << limit << std::endl;

        // show first MaxPages=10 pages:
        while (pageNotEmpty && page <= MaxPages) {
            ++page;
            std::cout << "> Page " << page << ":" << std::endl;
            pageNotEmpty = SelectPaging(client, path, limit, lastCity, lastNumber);
        }
    }
    catch (const TYdbErrorException& e) {
        std::cerr << "Execution failed due to fatal error: " << e.what() << std::endl;
        return false;
    }

    return true;
}
